### PR TITLE
feat(rating): added IMDB Radarr proxy

### PR DESCRIPTION
**Pull request by marcofaggian**
#### Description
Publishing upstream support for IMDB rating, added back RT audience score and cleaned API for ratings.

[Reference downstream in Jellyseerr](https://github.com/Fallenbagel/jellyseerr/pull/399)

#### Screenshots
immagine

### CODE REVIEW

**Comment by Fallenbagel** [comment]
[Screenshot with RT audience score]
image

File: server/api/rating/imdbRadarrProxy.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      nodeCache: cacheManager.getCache('imdb').data,
+    });
+  }
+
+  /**
+   * Search the RT algolia api for the movie title
+   *
+   * We compare the release date to make sure its the correct
+   * match. But it's not guaranteed to have results.
+   *
```
Comment (sct) [inline server/api/rating/imdbRadarrProxy.ts:172]: Can we update this comment for IMDB instead of RT?

File: server/api/rating/rottentomatoes.ts [hunk 2]
```diff
@@ -137,13 +137,17 @@ class RottenTomatoes extends ExternalAPI {
 
       return {
         title: movie.title,
+
```
Comment (sct) [inline server/api/rating/rottentomatoes.ts:140]: Lets get rid of this line break.

File: server/api/rating/imdbRadarrProxy.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      }
+
+      return {
+        title: data[0].Title,
+        url: `https://www.imdb.com/title/${data[0].ImdbId}`,
+        criticsScore: data[0].MovieRatings.Imdb.Value,
+      };
+    } catch (e) {
+      throw new Error(
+        `[RT API] Failed to retrieve movie ratings: ${e.message}`
```
Comment (sct) [inline server/api/rating/imdbRadarrProxy.ts:193]: We should probably change this from RT API.

File: server/api/rating/types.d.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { type IMDBRating } from '@server/api/rating/imdbRadarrProxy';
+import { type RTRating } from '@server/api/rating/rottentomatoes';
+
+export interface RatingResponse {
+  rt?: RTRating;
+  imdb?: IMDBRating;
+}
```
Comment (sct) [inline server/api/rating/types.d.ts:7]: Let's follow the same pattern as the other type files. Create a ratings.ts file in the root of `server/api` instead of a declaration file like this.

File: server/routes/movie.ts [hunk 5]
```diff
@@ -130,14 +133,24 @@ movieRoutes.get('/:id/ratings', async (req, res, next) => {
...
       });
     }
 
-    return res.status(200).json(rtratings);
+    const ratings: RatingResponse = {
+      ...(rtratings ? { rt: rtratings } : {}),
+      ...(imdbRatings ? { imdb: imdbRatings } : {}),
+    };
+
+    return res.status(200).json(ratings);
```
Comment (sct) [inline server/routes/movie.ts:153]: This is a breaking API change. We should create a new endpoint for this newly returned object shape.

File: server/api/rating/imdbRadarrProxy.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
+import ExternalAPI from '@server/api/externalapi';
+import cacheManager from '@server/lib/cache';
+
+export type IMDBRadarrProxyResponse = IMDBMovie[];
+
+export interface IMDBMovie {
```
Comment (sct) [inline server/api/rating/imdbRadarrProxy.ts:6]: Let's not export interfaces that are only used in this one file. (Same goes for other interfaces in here)

File: server/routes/movie.ts (hunk 5 above)
Comment (marcofaggian) [reply server/routes/movie.ts:153]: @sct  How do you envision this change?

My line of thought is: since the `ratings` returned previously were coupled to rt and here the object returned is less specific and more true to the generic path`/:id/ratings`, this is more like a path correction of something that should have been `/:id/ratings/rottentomatoes`.

But it is certainly a breaking change.

Comment (sct) [reply server/routes/movie.ts:153]: Can just name it something like `ratingscombined`. Just important we don't break the current endpoint.

Comment (marcofaggian) [reply server/routes/movie.ts:153]: Got it, pushed a new comment about it. I didn't see a script to update the openapi definition, should I update it manually?

Comment (sct) [reply server/routes/movie.ts:153]: Yes you will nee to update it manually

File: server/routes/movie.ts [hunk 7]
```diff
@@ -118,7 +118,7 @@ movieRoutes.get('/:id/similar', async (req, res, next) => {
   }
 });
 
-movieRoutes.get('/:id/ratings', async (req, res, next) => {
+movieRoutes.get('/:id/ratingscombined', async (req, res, next) => {
```
Comment (sct) [inline server/routes/movie.ts:121]: This is still a breaking change. We can't get rid of an endpoint either. Just make a new one entirely.

Comment (marcofaggian) [reply server/routes/movie.ts:162]: @sct reinstated the previous endpoint

File: overseerr-api.yml [hunk 8]
```diff
@@ -1,7 +1,7 @@
-openapi: '3.0.2'
+openapi: "3.0.2"
```
Comment (sct) [inline overseerr-api.yml:1]: We don't need changes to double quotes here. Can you revert these changes and only add your changes to the spec?

Comment (marcofaggian) [reply overseerr-api.yml:1]: Prettier doint its magic, commit coming

**Comment by marcofaggian** [comment]
@sct Is this one good to go?

**Comment by sct** [comment]
> @sct Is this one good to go?

No need to ping, haha. I know this is still here 😄 Busy weekend with the family. Will get to this next week.

**Comment by sct** [comment]
Please wait a little longer @marcofaggian we are just trying to get some bug fixes merged and released before we merge in any new features!

**Comment by marcofaggian** [comment]
> Please wait a little longer @marcofaggian we are just trying to get some bug fixes merged and released before we merge in any new features!

That's fine.

I was also thinking about developing this feature into something that can be configured in the onboarding/setup.
Something like "Choose your preferred rating providers", so that we could adopt more whilst avoiding to clutter the UI.

**Comment by Uruknara** [comment]
That's really an amazing idea !

**Comment by sct** [comment]
> I was also thinking about developing this feature into something that can be configured in the onboarding/setup. Something like "Choose your preferred rating providers", so that we could adopt more whilst avoiding to clutter the UI.

Not a bad idea. Definitely something for the normal settings pages. Maybe not setup. What other rating providers were you thinking about?

**Comment by marcofaggian** [comment]
I would have liked to add Metacritic and CommonSense Media, they seem widely used.
CommonSense requires a direct contact with them (I assume that their API is paywalled), and I haven't found a viable proxy.
For implementing Metacritic: I've found a [freemium proxy for metacritic](https://www.omdbapi.com/apikey.aspx) with a free tier that's generous and reasonable enough.

So in brief both would need an api key to work, no public proxy like IMDB's one.

Should I open a draft for a new PR about this settings improvements?

**Review by sct — APPROVED** [review]

**Comment by sct** [comment]
Sorry for the wait. We released the fixes we wanted to and can now do some feature merges again!

> I would have liked to add Metacritic and CommonSense Media, they seem widely used. CommonSense requires a direct contact with them (I assume that their API is paywalled), and I haven't found a viable proxy. For implementing Metacritic: I've found a [freemium proxy for metacritic](https://www.omdbapi.com/apikey.aspx) with a free tier that's generous and reasonable enough.
> 
> So in brief both would need an api key to work, no public proxy like IMDB's one.
> 
> Should I open a draft for a new PR about this settings improvements?

I have strayed from wanting to support these proxy APIs because of the fact they are somewhat limited until you pay. I think most OS instances would easily go over the 1000-a-day limit for OMDB. I suppose if enough people are interested, and some people want to pay the premium to get better ratings, it wouldn't hurt? I'm not against a draft PR but I don't want you to spend a ton of time for something we decide to not support. Maybe this is something we can further discuss in Discord?

**Comment by sct** [comment]
@all-contributors please add @marcofaggian for code
